### PR TITLE
addition of method_call_ctx_factory

### PR DESCRIPTION
**Pull request by asmirnov69**
method_call_ctx_factory global introduced to allow pandas_flavor extension in direction used by upcoming pyjviz public release. Also, version of pandas_flavor is set to 0.4.0

Default value of method_call_ctx_factory is None which is handled in easily verifiable way. None value causes no effect on pandas_flavor behavior. I.e. proposed release will have guaranteed no effect on pyjanitor if pandas_flavor 0.4.0 is installed

### CODE REVIEW

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py:140]: The new version of `__call__` has become more complex than before. Previously, we didn't have a docstring; while not optimal, it's still okay b/c of the relative simplicity of the original implementation. With the additional code, though, I think we should add a docstring. 

Here is something that I used Tabnine to generate as a starter:

```python
                """
                Call the method with the given arguments and return the result.

                This method is called when the user calls the method on the
                accessor object. It calls the underlying method with the
                arguments and keyword arguments passed to it.

                The method_call_context_factory is used
                to record the methods that were called.

                Parameters
                ----------
                *args : list
                    The arguments to pass to the method.
                **kwargs : dict
                    The keyword arguments to pass to the method.

                Returns
                -------
                object :
                    The result of calling the method.
                """
```

Would you be open to modifying it to describe the behaviour of `method_call_ctx_factory accurately`? (I will admit that I lost context (pun unintended) on pyjviz because of the myriad of activities happening at home, so recording it in-line here can be helpful.)

**Review by ericmjl — CHANGES_REQUESTED** [review]
Just one minor change request - to add documentation here please!

File: pandas_flavor/register.py
Comment (asmirnov69) [reply pandas_flavor/register.py:140]: will do, thanks for the feedback

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]: Function spelling change.

```suggestion
def handle_pandas_extension_call(method, method_signature, obj, args, kwargs):
```

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]:
```suggestion
    This function is called when the user calls a pandas DataFrame object's method.
```

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]:
```suggestion
    The pandas extension mechanism passes args and kwargs of original method call as it is applied to obj
```

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]: Improved grammar and switched to using Markdown syntax.

```suggestion
    Our implementation uses the global variable `method_call_ctx_factory`.
```

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]: I modified the grammar a bit here.

```suggestion
    `method_call_ctx_factory` can be either None or an abstract class.

    When `method_call_ctx_factory` is None, the implementation calls the registered method with unmodified args and kwargs and returns underlying method result.

    When `method_call_ctx_factory` is not None, `method_call_ctx_factory` is expected to refer to the function to create the context object. 
    The context object will be used to process inputs and outputs of `method` calls.
    It is also possible that the context object method `handle_start_method_call`
    will modify original args and kwargs before `method` call.
```

Comment (ericmjl) [inline pandas_flavor/register.py]: @asmirnov69 thinking from the perspective of a future maintainer, could you help include wording that describes where we would define `method_call_ctx_factory`? Given the state of the code as of now, it's unclear to me whether `pyjanitor` users need to define it or not. Also, if future maintainers need to modify `method_call_ctx_factory`, where and how would we go to modify it?

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]: Modified to use Markdown syntax.

```suggestion
    `method_call_ctx_factory` is a function that should have the following signature:
    
     `f(method_name: str, args: list, kwargs: dict) -> MethodCallCtx`

```

File: pandas_flavor/register.py
Comment (ericmjl) [inline pandas_flavor/register.py]: Minor change - switched to indented format to make it easier to read.

```suggestion
    MethodCallCtx is an abstract class:
    class MethodCallCtx(abc.ABC):

        @abstractmethod
        def __enter__(self) -> None: 
            raise NotImplemented

        @abstractmethod
        def __exit__(self, exc_type, exc_value, traceback) -> None: 
            raise NotImplemented

        @abstractmethod
        def handle_start_method_call(self, method_name: str, method_signature: inspect.Signature, method_args: list, method_kwargs: dict) -> tuple(list, dict): 
            raise NotImplemented

        @abstractmethod
        def handle_end_method_call(self, ret: object) -> None: 
            raise NotImplemented
```

**Review by ericmjl — CHANGES_REQUESTED** [review]
Many thanks @asmirnov69! I think we're getting close to making this PR perfect. 

I am requesting a few more changes. Most of them are implemented as suggested changes, which should make it easy for you to accept them on the GitHub PR interface, if you agree that they would be improvements on the current state. 

Upon re-review, I still felt a bit confused about `method_call_ctx_factory`. My primary question was, "Who defines this factory function?" If we could document that point, I think the documentation surrounding `method_call_ctx_factory` would be much clearer.

If there are no more major code or documentation changes, then I expect we should be able to merge asap!

**Comment by ericmjl** [comment]
As for pre-commit.ci failing, once you're done with your changes, if you don't mind I'll make a bunch of code changes to fix those issues raised by pre-commit.ci.

File: pandas_flavor/register.py
Comment (asmirnov69) [reply pandas_flavor/register.py]: method_call_ctx_factory is to allow pluggable module to be attached to pandas_flavor. I guess we may reconsider usage of global var as module interface and introduce plugin api. It could be justified if you envision existence of pandas_flavor as independent module in long term.

The only purpose of method_call_ctx_factory introduction is to allow pandas_flavor registered method call args and returns to be processed in plugin package. I was thinking about writing chapter into **pandas_flavor** README or other doc describing details. 

Thinking this way pyjanitor users would not need to be aware of underlying pandas_flavor module extensibility achieved via method_call_ctx_factory introduction. The value of method_call_ctx_factory suppose to be set by other package initialization (like pyjviz). Log message about method_call_ctx_factory value being set seems to be sufficient even for advanced users. If pyjanitor user decides not to use pyjviz then method_call_ctx_factory will stay None and all method call details logic will be bypassed.

**Comment by asmirnov69** [comment]
As I mentioned `method_call_ctx_factory` deserves to be described as contribution to pandas_flavor doc. I can put this as an pandas_flavor issue to track its progress separately after PR merge.

**Comment by ericmjl** [comment]
> As I mentioned `method_call_ctx_factory` deserves to be described as contribution to pandas_flavor doc. I can put this as an pandas_flavor issue to track its progress separately after PR merge.

@asmirnov69 I'm in agreement here! Okie dokes, let me take care of the pre-commit issues in a few pushes, and then I'll merge. We can cut a new release of `pandas_flavor` straight away after merging.

**Review by ericmjl — APPROVED** [review]
With all checks passing, let's merge and release!

**Comment by asmirnov69** [comment]
the issue was added about `method_call_ctx_factory` documentation to be written
https://github.com/pyjanitor-devs/pandas_flavor/issues/29

**Pull request merged**
